tools/python/openlr/quality.py

```diff
@@ -0,0 +1,143 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import xml.etree.ElementTree as ET
+import operator
+
+from itertools import islice
+from collections import namedtuple
+from math import sin, cos, atan2, radians
+
+LatLon = namedtuple('LatLon', 'lat, lon')
+
+def distance(x, y):
+    φ1, φ2 = map(radians, [x.lat, y.lat])
+    λ1, λ2 = map(radians, [x.lat, y.lat])
+    Δφ = φ2 - φ1
+    Δλ = λ2 - λ1
+    a = sin(Δφ/2)**2 + cos(φ1) * cos(φ2) * sin(Δλ/2)**2
+    return 2 * atan2(a**0.5, (1 - a)**0.5)
+
+def polyline_length(ps):
+    if len(ps) < 2:
+        return 0.0
+    return sum(
+        distance(ps[i], ps[i - 1])
+        for i in range(1, len(ps))
+    )
+
+def common_subpath(l1, l2, eq=operator.eq):
+    prefs_len = [
+        [0 for _ in range(len(l2) + 1)]
+        for _ in range(len(l1) + 1)
+    ]
+    for i in range(1, len(l1) + 1):
+        for j in range(1, len(l2) + 1):
+            if eq(l1[i - 1], l2[j - 1]):
+                prefs_len[i][j] = prefs_len[i - 1][j - 1] + 1
+    l, i, j = max((
+        (prefs_len[i][j], i, j)
+        for i in range(1, len(l1) + 1)
+        for j in range(1, len(l2) + 1)
+    ))
+    return i - l, j - l, l
+
+def common_part(l1, l2):
+    s1, s2, l = common_subpath(l1, l2)
+    intersection = polyline_length(l1[s1:s1 + l + 1])
+    difference = (
+        polyline_length(l1[:s1 + 1]) +
+        polyline_length(l2[:s2 + 1]) +
+        polyline_length(l1[s1 + l - 1:]) +
+        polyline_length(l2[s2 + l - 1:])
+    )
+    total_length = intersection + difference
```

Comment:
Total length of what? Shouldn't you count intersection twice here?

Reply:
2 * intersection does not work well for us.
Done by inlining intersection / (intersection + difference).